resolwe_bio/processes/support_processors/merge_samples.py

```diff
@@ -1,218 +1,190 @@
     category = "Other"
     scheduling_class = SchedulingClass.BATCH
-    entity = {
-        "type": "sample",
-        "descriptor_schema": "sample",
-        "always_create": True,
-    }
     requirements = {
         "expression-engine": "jinja",
         "executor": {
             "docker": {"image": "public.ecr.aws/s4q6j6e8/resolwebio/common:2.3.1"}
         },
+        "relations": [{"type": "group"}],
     }
-    data_name = '{{ reads|map("sample_name")|join(", ")|default("?") }}'
+    data_name = "Merge FASTQ (single-end)"
 
     class Input:
         """Input fields to process MergeFastqSingle."""
 
         reads = ListField(
             DataField(data_type="reads:fastq:single:"),
-            label="Reads data objects",
+            label="Select relations",
+            description="Define and select replicate relations.",
+            relation_type="group",
         )
 
-    class Output:
-        """Output fields to process MergeFastqSingle."""
-
-        fastq = ListField(FileField(), label="Reads file")
-        fastqc_url = ListField(
-            FileHtmlField(),
-            label="Quality control with FastQC",
-        )
-        fastqc_archive = ListField(FileField(), label="Download FastQC archive")
-
     def run(self, inputs, outputs):
         """Run the analysis."""
-        name = f"{Path(inputs.reads[0].output.fastq[0].path).name.replace('.fastq.gz', '')}_merged"
-        merged_fastq = f"{name}.fastq.gz"
-        with open(merged_fastq, "wb") as outfile:
-            for reads in inputs.reads:
-                for fastq in reads.output.fastq:
-                    with open(fastq.path, "rb") as infile:
-                        for line in infile:
-                            outfile.write(line)
-
-        outputs.fastq = [merged_fastq]
-
-        print("Postprocessing FastQC...")
-        stderr = run_fastqc([merged_fastq], "./fastqc")
-        # FastQC writes both progress and errors to stderr and exits with code 0.
-        # Catch if file is empty, wrong format... (Failed to process) or
-        # if file path does not exist, file cannot be read (Skipping).
-        if "Failed to process" in stderr or "Skipping" in stderr:
-            self.error("Failed while processing with FastQC.")
-
-        fastqc = f"{name}_fastqc.zip"
-        fastqc_path = Path("fastqc")
-        move(str(fastqc_path / fastqc), ".")
-
-        report_dir = fastqc_path / f"{name}_fastqc"
-        fastqc_url = [
-            {
-                "file": str(report_dir / "fastqc_report.html"),
-                "refs": [str(fastqc_path)],
-            }
-        ]
-
-        outputs.fastqc_url = fastqc_url
-        outputs.fastqc_archive = [fastqc]
+
+        # Check if user has selected multiple read objects from the same sample
+        data_by_sample = {}
+        for data in inputs.reads:
+            if data.entity_id in data_by_sample:
+                self.warning(
+                    "There are multiple read objects for "
```

Comment:
It is common for the sample to have trimmed reads objects attached. For any subsequent joining of already analysed data, the original input (FASTQ) that originated the entity might make more sense. So, I propose to merge the first FASTQ on the sample, not the last.

Reply:
Thanks that is a good suggestion.